### PR TITLE
Fix AST rendering (many cases) and parsing (a few) (fixes #430)

### LINKED ISSUE #430: sql: edge cases in parser
Scalacheck tests of unparsing-parsing roundtrips turned up a handful of edge cases where either the parser cannot handle part of the AST, or the `sql` method does not properly encode things.

See comments in SqlParserSpec.scala.

### PR DESCRIPTION
**Pull request by mossprescott**
Parsing fixes:

1. Accept parens around joins.
2. Allow arbitrary expressions inside `[]` and `{}`.
3. Parse numeric literals with exponents.
4. Rip out half-implemented
EXTRACT syntax.

Rendering:

1. Always surround joins with parens.
2. Escape single quotes in string
constants.
3. Quote identifiers as needed and escape double quotes in them.

This means `day` and `hour`, etc. are no longer keywords, at least until we
decide to support the awful EXTRACT syntax (see #210).

There were some parsing tests that included EXTRACT, which I just modified to
use `date_part`. As long as we have an equivalent that works, I didn't see any
point in preserving a test for the syntax we've never supported in the
compiler.

### CODE REVIEW

File: core/src/main/scala/slamdata/engine/sql/parser.scala [hunk 1]
```diff
@@ -287,9 +279,11 @@ class SQLParser extends StandardTokenParsers {
     ident ~ opt(keyword("as")) ~ opt(ident) ^^ {
       case ident ~ _ ~ alias => TableRelationAST(ident, alias)
     } |
-    op("(") ~ select ~ op(")") ~ opt(keyword("as")) ~ ident ^^ {
-      case _ ~ select ~ _ ~ _ ~ alias => SubqueryRelationAST(select, alias)
-    }
+    op("(") ~> (
```
Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/sql/parser.scala]: This doesn't quite do it, as it doesn't balance the parens. You can factor out the middle part and repeat it twice, once for the parens (list first), once without them.

Comment (jdegoes) [inline core/src/main/scala/slamdata/engine/sql/parser.scala]: Oops, disregard that, was reading that as "optional paren" instead of "paren op". :smile:

File: core/src/test/scala/slamdata/engine/sql/SqlParserSpec.scala [hunk 2]
```diff
@@ -223,32 +223,39 @@ class SQLParserSpec extends Specification with ScalaCheck with DisjunctionMatche
...
-      10 -> simple,
+      5 -> simple,
       1 -> (for {
         s <- selectGen(2)
         c <- Gen.alphaChar
       } yield SubqueryRelationAST(s, c.toString)),
       1 -> (for {
-        l <- relationGen(0)  // Note: we do not parenthesize nested joins, and the parser get confused
-        r <- relationGen(0)
+        l <- relationGen(depth-1)
```
Comment (jdegoes) [inline core/src/test/scala/slamdata/engine/sql/SqlParserSpec.scala]: Remind me sometime to show you the alternate way to encode this using sized generators (advantage being it passes the parameter around for you).

**Comment by jdegoes** [comment]
:shipit:
